### ISSUE TITLE
Recursive upgrading project dependencies

### ISSUE THREAD

**Issue by dmitry-kazakov** [alire-project/alire#237]
The objective is walking the dependencies list and switching to the latest available version of each keeping the list consistent. Example. Let be 4 projects. A, B, C, D. Bv1 depends on Av1. Cv1 depends on Av1. Dv1 depends on Bv1 and Cv1. Now there is Av2 and Bv4 dependent on Av2 etc. The objective is to upgrade D to the latest versions of A, B, C. Options would be upgrading C to Av2 or rejecting upgrade of D since there is no version of C working with Av2.

**Comment by mosteo** [comment]
If I understand what you mean correctly, currently there is no management of individual versions: whenever the list of dependencies is needed (`alr get`, `alr build`, probably in some other cases I'm missing right now) the solver is invoked and it will find a consistent solution, if there is one. The current solver is exhaustive and will attempt all combinations in the worst case, so a dependency might be resolved as a lower version than available if another dependent requires it (as in your example IIUC). 

Dependencies are attempted by default from newest to oldest and the objective is to find a consistent solution, not individually managing the dependencies (e.g. I want to only upgrade C is not directly supported right now). The latter would be a new use case.

Currently the solver is invoked every time the list of dependencies is needed. I had in the radar the idea of caching this resolution so the solver is only called when explicit upgrades are requested. This cached list would make easy to upgrade only a subset of dependencies, considering the rest as "pinned".

But, I'm not sure if that's what you want. Perhaps what you want is already happening with:

1. `alr index --update-all`
1. `alr update`

The first retrieves new crate versions, and the second call will re-resolve dependencies, upgrading whatever can be upgraded. Right now, to have finer control of versions, one has to `alr pin`, which will freeze all versions in place, and manually edit `<crate>/alire/crate_name.toml` to allow different version sets therein.

**Comment by dmitry-kazakov** [comment]
Then it is another issue. There should be possible to pull a certain version of a package with all dependencies corresponding to that version. Pulling the latest versions is inconsistent. A version of a package must always refer to a definite version of the packages it depends on. It cannot be open-ended. Only version equality should be allowed. So, let the package B depend on A. Then it is always like Bv1->Av2, Bv2->Av5, Bv3- Av5, Bv4->Av7 etc. When Bv2 is pulled, then Av5 pulled with it, not Av7.

**Comment by mosteo** [comment]
This is doable by specifying exact versions in the release dependencies. The option to use version sets as dependencies is up to the publisher of a release.
